src/client/src/app/site/[site]/details/edit-reference-details/edit-reference-details-form.tsx

```
@@ -0,0 +1,129 @@
+  wellKnownOdsCodeEntries: WellKnownOdsEntry[];
+}) => {
+  const {
+    register,
+    handleSubmit,
+    formState: { errors, isSubmitting, isSubmitSuccessful },
+  } = useForm<FormFields>({
+    defaultValues: {
+      odsCode: siteWithAttributes.odsCode,
+      icb: siteWithAttributes.integratedCareBoard,
+      region: siteWithAttributes.region,
+    },
+  });
+
+  const mapWellKnownCodes = (type: string): SelectOption[] =>
+    wellKnownOdsCodeEntries
+      .filter(e => e.type === type)
+      .map(x => {
+        return {
+          value: x.odsCode,
+          label: x.displayName,
+        };
+      });
+
+  const validateWellKnownCode = (value: string, type: string): boolean => {
+    const wellKnownICB = wellKnownOdsCodeEntries.find(
+      x => x.type === type && x.odsCode === value,
+    );
+
+    return wellKnownICB !== undefined;
+  };
+
+  const { replace } = useRouter();
+
+  const submitForm: SubmitHandler<FormFields> = async (form: FormFields) => {
+    const payload: SetSiteReferenceDetailsRequest = {
+      odsCode: form.odsCode.trim(),
+      icb: form.icb,
+      region: form.region,
+    };
+    await saveSiteReferenceDetails(siteWithAttributes.id, payload);
+
+    replace(`/site/${siteWithAttributes.id}/details`);
+  };
+
+  return (
+    <form onSubmit={handleSubmit(submitForm)}>
+      <FormGroup
+        error={errors.odsCode ? 'You have not entered an ODS code' : undefined}
+      >
+        <TextInput
+          id="odsCode"
+          label="ODS code"
+          {...register('odsCode', {
+            required: true,
+          })}
+        ></TextInput>
+      </FormGroup>
+      <FormGroup
+        error={errors.icb ? 'You have not selected an ICB' : undefined}
+      >
+        <Select
+          id="icb"
+          label="ICB"
+          options={mapWellKnownCodes('icb')}
+          {...register('icb', {
+            required: true,
+            validate: e => validateWellKnownCode(e, 'icb'),
```

I'm all for the principle of failing fast, but this doesn't feel like it should be necessary. We've just set `options` to a map of this object so should know that their values are correct. I presume the backend has its own validation on this too if an erroneous value did get through

---

If we don't trust what's in the db but we DO know how to validate it, we're then in a world where we should be writing validators for retrieval and validating the response from each GetFooFunction.cs. 

Obviously this is a huge red flag if we do this, but it's better than passing the problem along to the frontend

---

To be fair it sounds like there is a ticket for validating the MYA Bulk Data now, so I'm not really sure where this leaves my work in terms of validation. Might need to chat to some BA's about it.